### ISSUE TITLE
Things to discuss

### ISSUE THREAD

**Issue by dpc** [stevenroose/rust-bitcoindrpc#3]
I'm just going to write a paragraph for each item to discuss. Please don't hesitate to tell me when you disagree.

I am (or at least used to be) a tab-person myself and cried for 5 minutes when it was decided that Rust will go with spaces, but since then got used to it, and I believe it's better to go with the ecosystem and have all the code feel the same, than not. So can we switch to spaces?

I've implemented the no-alloc macro that we've talked about. Please see `master` and tell me what you think.

I feel like, since we're exposing a Rust API, should name the methods like `get_raw_transaction`. It's easier to read and just feel right.

If we have a method `get_raw_transaction`, we should just return `String` with the hex-encoded data, since that's exactly what the node returns. If the user wants `bitcoin` crate type, we should have the `get_by_id`/`Querable` methods for this, or if needed additional methods, that are named somewhat differently than the standard calls.

**Comment by dpc** [comment]
We should use hex decoding from `bitcoin` itself, just to save on one dependency.

**Comment by dpc** [comment]
I don't think `jsonrpc::Client` needs mutability, so we probably don't need to make all methods `&mut self`. Unless there's a chance that `jsonrpc::Client` will change.

**Comment by dpc** [comment]
`bitcoindrpc` crate name... As per https://github.com/rust-bitcoin/rust-bitcoin/issues/184#issuecomment-434742247 , should we rename it to ... `bitcoincore-rpc` or something? I specifically don't like the the fact that there's no `-` or `_` in the name, which how usually crates are named.

**Comment by dpc** [comment]
Is `failure` in `[dev-dependencies]` ok?

**Comment by dpc** [comment]
I've ported a simple example that I was planing to use a real-world test against a real full node.

```
cargo run --example test_against_node http://localhost:8332 user password
```

**Comment by stevenroose** [comment]
> We should use hex decoding from `bitcoin` itself, just to save on one dependency.

One thing I don't like about this is that users that use the `hex` crate and this crate suddenly have two hex errors in their scope.

**Comment by stevenroose** [comment]
Can we rename `Querable` to [`Queryable`](https://www.reddit.com/r/AskReddit/comments/7i4vp/askreddit_is_queryable_a_proper_english_word_any/c06pxu8/) and put it in a separate file. Also, `json::GetTxOutResult` can be queryable by `bitcoin::OutPoint`.

**Comment by dpc** [comment]
> What happens with blocks that can be queried by both numbers and hashes? :)


We could parametrize `Querable` over both T and Id, so one can query tx with get_by_id(blockhash) or `get_by_id((blockhash, txhash))`, and blocks with both `get_by_id(height)`, `get_by_id(hash)`.

I would probably rename it to `get_by`, since `height` is not an Id?

Is it worth it though? I imagine, in most cases, one just wants to query by hash. And querrying by height is risky, since there's always a chance there was a reorg, and underlying block changed. But I guess that's the user problem to be aware of that.

> Does that expect testnet or mainnet node specifically? Or is it generic?

I tested mainnet. But a testnet option could be added.

> One thing I don't like about this is that users that use the hex crate and this crate suddenly have two hex errors in their scope.

IMO, exposing  all these variants is pointless in the first place. I can't imagine someone actually matching them and doing something useful with them. The only useful thing seems the human-readable message about what was the error about.

> Can we rename Querable to Queryable and put it in a separate file. Also, json::GetTxOutResult can be queryable by bitcoin::OutPoint.

Sure. I make a lot of naming mistakes, so feel free to correct me at will. I'm doing my best. :D

**Comment by dpc** [comment]
I think taking `&mut self` might be needed. It depends how is your change to make `rust-jsonrpc` have a custom backend will be implemented.

**Comment by dpc** [comment]
I've nuked the macros.

**Comment by stevenroose** [comment]
> I think taking `&mut self` might be needed. It depends how is your change to make `rust-jsonrpc` have a custom backend will be implemented.

Hmm, true. Some clients might need it, others don't.

**Comment by dpc** [comment]
> Hmm, true. Some clients might need it, others don't.

Yes. So once the dust settles in `jsonrpc` we should decide who is responsible for thread-safety/mutability, or should `Client` be just taken by `&mut`. I am guessing it's not a big deal one way or the other.

**Comment by stevenroose** [comment]
I think we should be quite release-ready once jsonrpc lands a release and then we depend on the latest there. Even though in principle, I think we can support older versions of jsonrpc as well, no?

**Comment by dpc** [comment]
Since this crate is still in 0.x we can release whenever. No need to even wait for jsonrpc release. We can release once now, and then afterwards.
